juniper_hyper/src/lib.rs

```diff
@@ -20,48 +21,133 @@ use std::error::Error as StdError;
+            future::done(
                 request
                     .uri()
                     .query()
-                    .map(gql_request_from_get)
+                    .map(|q| gql_request_from_get(q).map(GraphQLRequest::Single))
                     .unwrap_or(Err(GraphQLRequestError::Invalid(
                         "'query' parameter is missing".to_string(),
                     ))),
-            ).map(|res| new_response(StatusCode::METHOD_NOT_ALLOWED)),
+            ).and_then(move |gql_req| execute_request(pool, root_node, context, gql_req))
+            .or_else(|err| future::ok(render_error(err))),
         ),
         &Method::POST => Box::new(
             request
                 .into_body()
                 .concat2()
-                .map(move |chunk| {
-                    String::from_utf8(chunk.iter().rev().cloned().collect::<Vec<u8>>())
-                        .map_err(GraphQLRequestError::BodyUtf8)
-                        .and_then(|input| {
-                            serde_json::from_str::<GraphQLRequest>(&input)
-                                .map_err(GraphQLRequestError::BodyJSONError)
-                        })
-                }).or_else(|err| future::ok(Err(GraphQLRequestError::BodyHyper(err))))
-                .map(|res| new_response(StatusCode::METHOD_NOT_ALLOWED)),
+                .or_else(|err| future::done(Err(GraphQLRequestError::BodyHyper(err))))
+                .and_then(move |chunk| {
+                    future::done({
+                        String::from_utf8(chunk.iter().cloned().collect::<Vec<u8>>())
+                            .map_err(GraphQLRequestError::BodyUtf8)
+                            .and_then(|input| {
+                                serde_json::from_str::<GraphQLRequest>(&input)
+                                    .map_err(GraphQLRequestError::BodyJSONError)
+                            })
+                    })
+                }).and_then(move |gql_req| execute_request(pool, root_node, context, gql_req))
+                .or_else(|err| future::ok(render_error(err))),
         ),
         _ => return Box::new(future::ok(new_response(StatusCode::METHOD_NOT_ALLOWED))),
     }
 }
 
+fn render_error(err: GraphQLRequestError) -> Response<Body> {
+    let message = format!("{}", err);
+    let mut resp = new_response(StatusCode::UNPROCESSABLE_ENTITY);
+    *resp.body_mut() = Body::from(message);
+    resp
+}
+
+fn execute_request<CtxT, QueryT, MutationT, Err>(
+    pool: CpuPool,
+    root_node: Arc<RootNode<'static, QueryT, MutationT>>,
+    context: Arc<CtxT>,
+    request: GraphQLRequest,
+) -> impl Future<Item = Response<Body>, Error = Err>
+where
+    CtxT: Send + Sync + 'static,
+    QueryT: GraphQLType<Context = CtxT> + Send + Sync + 'static,
+    MutationT: GraphQLType<Context = CtxT> + Send + Sync + 'static,
+    QueryT::TypeInfo: Send + Sync,
+    MutationT::TypeInfo: Send + Sync,
+    Err: Send + Sync + 'static,
+{
+    pool.spawn_fn(move || {
```

Comment:
Doesn't this technically violate the [spec guarantee](https://graphql.org/learn/queries/#multiple-fields-in-mutations) that mutation operations are always executed sequentially?

Reply:
This is spawning a whole GraphQL request on the pool, so the mutations are not executed in parallel (that part is handled by Juniper of course).